### PR TITLE
Kristen Tortorelli NDVI homework review

### PR DESCRIPTION
**Pull request by kristentortorelli**
# Pull Request Template

## CI Checks
- [ ] The notebook runs from start to finish on all operating systems:
   - [ ] Mac
   - [ ] Windows
   - [ ] Linux 
### Reproducibility
- [ ] Are the data downloaded in the code
- [ ] Are paths created to ensure they work on all operating systems (using  os.path.join)
- [ ] Are comments used to clarify the contents of the code that can’t be clarified using expressive variable and function names alone? (not too many comments - just enough)
- [ ] Does the notebook run from start to finish?

### Functions
- [ ] Do functions follow PEP 8 format conventions?
- [ ] Are function docstrings clear (all  inputs and outputs clearly  described and defined)
- [ ] Are function names expressive (the name describes what the function does)?
- [ ] Are functions easy to understand and read?
- [ ] How many tasks does each function do? (ideally a function does one thing well).
### Package imports
- [ ] Are standard modules (those included with the base  python install) vs. third party (related but externally developed tools) import groups correct with appropriate spacing in between each group?
- [ ] Are variable names throughout the code, expressive?
- [ ] Suggest changes if not, highlight what was done well
- [ ] Is the code overall easy to understand and read? Are there things that would make it more clear / cleaner?

### DRY Code
- [ ] Are segments of code repeated in the file or is the code DRY?
- [ ] Are loops used to optimize DRY code?
- [ ] Are functions used to optimize DRY code?
- [ ] Are there any areas that could be potentially improved (you can suggest improvements OR you can just highlight parts of the code where you suspect it could  be cleaner / more efficient.

### Novel Approaches to Problem solving
- [ ] Highlight any novel approaches to completing the assignment.

### CODE REVIEW

File: ea-2022-04-ndvi-automation.ipynb [hunk 1]
```diff
@@ -0,0 +1,1600 @@
...
+    "import seaborn as sns\n",
+    "import numpy as np\n",
+    "from numpy import ma\n",
+    "import xarray as xr\n",
+    "import rioxarray as rxr\n",
+    "import earthpy as et\n",
+    "import earthpy.plot as ep\n",
+    "import earthpy.mask as em\n",
+    "import earthpy.spatial as es\n",
+    "\n",
```
Comment (ericrnutt5) [inline ea-2022-04-ndvi-automation.ipynb:262]: Looks like all of the necessary modules are here and used! Nice work!

File: ea-2022-04-ndvi-automation.ipynb [hunk 2]
```diff
@@ -0,0 +1,1600 @@
...
+    "date_single = '20170317'\n",
+    "datetime_object_single = [datetime.strptime(date_single, '%Y%m%d')]\n",
+    "\n",
+    "df_harv_single_date = pd.DataFrame(\n",
+    "    {'mean_ndvi': ndvi_values}, index=datetime_object_single)\n",
+    "df_harv_single_date['site'] = 'HARV'\n",
+    "df_harv_single_date"
+   ]
+  },
+  {
```
Comment (ericrnutt5) [inline ea-2022-04-ndvi-automation.ipynb:728]: Love all of your commented steps. It makes it very easy to read and work through. Aids in reproducibility!

File: ea-2022-04-ndvi-automation.ipynb [hunk 3]
```diff
@@ -0,0 +1,1600 @@
...
+    "\n",
+    "# Set consistent plotting style\n",
+    "sns.set_style(\"white\")\n",
+    "sns.set(font_scale=1.5)\n",
+    "\n",
+    "# Download the data\n",
+    "data = et.data.get_data('ndvi-automation')\n",
+    "\n",
+    "# Create variable for data path\n",
+    "data_path = os.path.join(et.io.HOME, 'earth-analytics', 'data')\n",
```
Comment (ericrnutt5) [inline ea-2022-04-ndvi-automation.ipynb:271]: Great work using os.path.join to insure the notebook runs on all systems.

File: ea-2022-04-ndvi-automation.ipynb [hunk 4]
```diff
@@ -0,0 +1,1600 @@
...
+    "        Path to directory with all landsat scenes/date folders to be combined.\n",
+    "    crop_path: string\n",
+    "        A string path to the geodataframe containing the crop geometry that \n",
+    "        you wish to crop your data to.\n",
+    "\n",
+    "    Returns\n",
+    "    ----------\n",
+    "    df : pandas dataframe\n",
+    "        Dataframe containing average NDVI values, date, and site name. \n",
+    "        \"\"\"\n",
```
Comment (ericrnutt5) [inline ea-2022-04-ndvi-automation.ipynb:569]: Great pseudocode for all of your functions. Easy to read and understand what each function does!

File: ea-2022-04-ndvi-automation.ipynb [hunk 5]
```diff
@@ -0,0 +1,1600 @@
...
+    "\n",
+    "    return df"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Using Functions for Optimized Code \n",
+    "These functions defined above help to make my notebook code more efficient and optimized. Instead of repeating several lines of code for every time I need to calculate NDVI, open/clean/crop a list of raster files, remove cloud interference from landsat data, or combine NDVI values for several sites/scenes into a pandas dataframe, I can use the functions defined above. These functions reduce the potential for error, make it easier to add more data to this analysis, and help someone else understand my code better. "
```
Comment (ericrnutt5) [inline ea-2022-04-ndvi-automation.ipynb:606]: Great use of multiple functions to complete the workflow! It is possible to combine some steps in the above functions into two functions to complete the workflow using for loops, but using multiple functions not only properly completes the required tasks, but adds an element of customization to future workflows that would not necessarily be possible with fewer functions. Great job!

File: ea-2022-04-ndvi-automation.ipynb [hunk 6]
```diff
@@ -0,0 +1,1600 @@
...
+    "\n",
+    "for site_name, group in mean_ndvi_df_total.groupby('site'):\n",
+    "    group.groupby('site').plot(y='mean_ndvi',\n",
+    "                               label=site_name,\n",
+    "                               linewidth=3.0,\n",
+    "                               ax=ax,\n",
+    "                               alpha=.8)\n",
+    "\n",
+    "# Set plot title and axes labels\n",
+    "ax.set(title=\"Mean NDVI Values from Landsat Data for Two Sites (Clouds Removed)\\n \\\n",
```
Comment (ericrnutt5) [inline ea-2022-04-ndvi-automation.ipynb:1280]: While by no means necessary, consider adding the (fontweight='bold') parameter to the figure subtitles for a more aesthetic plot. Again this is by no means a requirement, just a personal preference, so feel free to consider it in the future!

File: ea-2022-04-ndvi-automation.ipynb [hunk 7]
```diff
@@ -0,0 +1,1600 @@
...
+    "                March - December 2017\",\n",
+    "       xlabel=\"Date\",\n",
+    "       ylabel=\"NDVI\")\n",
```
Comment (ericrnutt5) [inline ea-2022-04-ndvi-automation.ipynb:1283]: Specify mean ndvi here?

**Review by ericrnutt5 — COMMENTED** [review]
@kristentortorelli I have completed my review! Sorry for the delay! Loved your use of multiple functions and really great overall code!

Let me know if you have any questions!

File: ea-2022-04-ndvi-automation.ipynb (hunk 1 above)
Comment (kristentortorelli) [reply ea-2022-04-ndvi-automation.ipynb:262]: Thank you!

File: ea-2022-04-ndvi-automation.ipynb (hunk 5 above)
Comment (kristentortorelli) [reply ea-2022-04-ndvi-automation.ipynb:606]: Thank you! I actually like the readability of your method (using fewer, more comprehensive functions with for loops) better. But this method did help me break the process down into smaller, more digestible pieces, which is what I needed in order to understand how to analyze data for all the sites/dates.

File: ea-2022-04-ndvi-automation.ipynb (hunk 6 above)
Comment (kristentortorelli) [reply ea-2022-04-ndvi-automation.ipynb:1280]: Yes - Thank you! I definitely will use in future notebooks, it helps the overall plot formatting a lot.

File: ea-2022-04-ndvi-automation.ipynb (hunk 7 above)
Comment (kristentortorelli) [reply ea-2022-04-ndvi-automation.ipynb:1283]: Oops - good call! I'm going to fix that to say mean NDVI. Thank you!

File: ea-2022-04-ndvi-automation.ipynb (hunk 4 above)
Comment (kristentortorelli) [reply ea-2022-04-ndvi-automation.ipynb:569]: Thank you!

File: ea-2022-04-ndvi-automation.ipynb (hunk 2 above)
Comment (kristentortorelli) [reply ea-2022-04-ndvi-automation.ipynb:728]: Thanks!

**Pull request closed without merging**